File: internal/mapfile/mapfile.go
package mapfile

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"kimen/internal/projection"
)

// Format (v0) is line-oriented and comment-friendly.
//
// Examples:
//
//	env LINJE_API_TOKEN=linje.prod.api_token
//	env PORT=const:5050
//	file key.json=gcp.sa_key_json
//	stdin exec:some-command
//	envpath GOOGLE_APPLICATION_CREDENTIALS=key.json
//
// Blank lines and lines starting with # are ignored.
//
// Rules:
// - "env" and "file" use the same mapping shape as flags: <lhs>=<value>.
// - "stdin" sets the command stdin source for run projections.
// - "envpath" maps an env var to a file path under KIMEN_FILES_DIR. It does not itself create the file mapping.
type Map struct {
	Request  projection.Request
	EnvPaths []projection.EnvPathMapping
}

func ParseFile(path string) (Map, error) {
	f, err := os.Open(path)
	if err != nil {
		return Map{}, err
	}
	defer f.Close()

	m, err := Parse(f)
	if err != nil {
		return Map{}, err
	}
	return m, nil
}

func Parse(r io.Reader) (Map, error) {
	var envMappings []string
	var fileMappings []string
	var envPathMappings []string
	var stdin string

	sc := bufio.NewScanner(r)
	lineNo := 0
	for sc.Scan() {
		lineNo++
		line := strings.TrimSpace(sc.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		// Strip inline comments: "…  # comment"
		if i := strings.Index(line, "#"); i > 0 {
			if strings.TrimSpace(line[:i]) != "" && (line[i-1] == ' ' || line[i-1] == '\t') {
				line = strings.TrimSpace(line[:i])
			}
		}

		kind, rest := splitKindAndRest(line)
		if kind == "" || rest == "" {
			return Map{}, fmt.Errorf("invalid map line %d: expected '<kind> <mapping>'", lineNo)
		}
		switch kind {
		case "env":
			envMappings = append(envMappings, rest)
		case "file":
			fileMappings = append(fileMappings, rest)
		case "stdin":
			if stdin != "" {
				return Map{}, fmt.Errorf("invalid map line %d: multiple stdin entries", lineNo)
			}
			stdin = rest
		case "envpath":
			envPathMappings = append(envPathMappings, rest)
		default:
			return Map{}, fmt.Errorf("invalid map line %d: unknown kind %q", lineNo, kind)
		}
	}
	if err := sc.Err(); err != nil {
		return Map{}, err
	}

	req, err := projection.ParseRequest(envMappings, fileMappings, stdin)
	if err != nil {
		return Map{}, err
	}
	envPaths, err := projection.ParseEnvPathMappings(envPathMappings)
	if err != nil {
		return Map{}, err
	}
	return Map{Request: req, EnvPaths: envPaths}, nil
}

func splitKindAndRest(line string) (kind, rest string) {
	for i, r := range line {
		if r == ' ' || r == '\t' {
			kind = line[:i]
			rest = strings.TrimSpace(line[i:])
			return kind, rest
		}
	}
	return "", ""
}

// ResolveProfile looks up a profile name to a map file path.
//
// Search order:
// - $KIMEN_PROFILE_DIR/<name>.kmap
// - ./.kimen/profiles/<name>.kmap (relative to cwd)
// - <UserConfigDir>/kimen/profiles/<name>.kmap
const envProfileDir = "KIMEN_PROFILE_DIR"

var profileNameRE = regexp.MustCompile(`^[A-Za-z0-9_.-]+$`)

func ResolveProfile(name string) (string, error) {
	name = strings.TrimSpace(name)
	if name == "" {
		return "", errors.New("empty profile name")
	}
	if !profileNameRE.MatchString(name) {
		return "", fmt.Errorf("invalid profile name %q (allowed: letters, digits, ., _, -)", name)
	}
	filename := name + ".kmap"

	if dir := os.Getenv(envProfileDir); dir != "" {
		p := filepath.Join(dir, filename)
		if _, err := os.Stat(p); err == nil {
			return p, nil
		}
	}

	p := filepath.Join(".kimen", "profiles", filename)
	if _, err := os.Stat(p); err == nil {
		return p, nil
	}

	cfgDir, err := os.UserConfigDir()
	if err != nil {
		return "", err
	}
	if cfgDir != "" {
		p = filepath.Join(cfgDir, "kimen", "profiles", filename)
		if _, err := os.Stat(p); err == nil {
			return p, nil
		}
	}

	return "", fmt.Errorf("profile %q not found (set %s or create .kimen/profiles/%s)", name, envProfileDir, filename)
}

File: src/kimen/mapfile.clj
(ns kimen.mapfile
  (:require
    [clojure.java.io :as io]
    [clojure.string :as str]))

(def env-var-re #"^[A-Za-z_][A-Za-z0-9_]*$")
(def profile-name-re #"^[A-Za-z0-9_.-]+$")
(def env-profile-dir "KIMEN_PROFILE_DIR")

(defn- invalid!
  ([msg]
   (throw (ex-info msg {:type :invalid-map})))
  ([msg data]
   (throw (ex-info msg (merge {:type :invalid-map} data)))))

(defn valid-profile-name?
  [s]
  (boolean (and (string? s) (re-matches profile-name-re s))))

(defn- normalize-relative-path
  [raw]
  (let [input (some-> raw str str/trim)]
    (when (str/blank? input)
      (invalid! "empty relative path"))
    (let [path (.normalize (java.nio.file.Paths/get input (make-array String 0)))
          clean (.toString path)]
      (when (.isAbsolute path)
        (invalid! (format "invalid relative path %s" (pr-str raw)) {:reason :invalid-relative-path}))
      (when (or (= clean ".")
                (= clean "..")
                (str/starts-with? clean "../")
                (str/starts-with? clean "..\\")
                (str/starts-with? clean "/")
                (str/starts-with? clean "\\"))
        (invalid! (format "invalid relative path %s" (pr-str raw)) {:reason :invalid-relative-path}))
      (str/replace clean "\\" "/"))))

(defn parse-request
  [env-mappings file-mappings stdin-spec]
  (let [envs
        (mapv (fn [m]
                (let [[var value] (str/split (str m) #"=" 2)
                      var (some-> var str/trim)
                      value (some-> value str/trim)]
                  (when-not (and var value)
                    (invalid! (format "invalid --env mapping %s (expected VAR=<value>)" (pr-str m))
                              {:reason :invalid-mapping}))
                  (when-not (re-matches env-var-re var)
                    (invalid! (format "invalid env var name %s" (pr-str var)) {:reason :invalid-env-var}))
                  (when (str/blank? value)
                    (invalid! "empty value in --env mapping" {:reason :invalid-mapping}))
                  {:var var :name value}))
              env-mappings)
        files
        (mapv (fn [m]
                (let [[rel value] (str/split (str m) #"=" 2)
                      rel (some-> rel str/trim)
                      value (some-> value str/trim)]
                  (when-not (and rel value)
                    (invalid! (format "invalid --file mapping %s (expected relpath=<value>)" (pr-str m))
                              {:reason :invalid-mapping}))
                  (when (str/blank? value)
                    (invalid! "empty value in --file mapping" {:reason :invalid-mapping}))
                  {:rel-path (normalize-relative-path rel)
                   :name value}))
              file-mappings)
        stdin-spec (some-> stdin-spec str str/trim)]
    {:envs envs
     :files files
     :stdin (when-not (str/blank? stdin-spec) stdin-spec)}))

(defn parse-envpath-mappings
  [envpath-mappings]
  (mapv (fn [m]
          (let [[var rel] (str/split (str m) #"=" 2)
                var (some-> var str/trim)
                rel (some-> rel str/trim)]
            (when-not (and var rel)
              (invalid! (format "invalid envpath mapping %s (expected VAR=relpath)" (pr-str m))
                        {:reason :invalid-mapping}))
            (when-not (re-matches env-var-re var)
              (invalid! (format "invalid env var name %s" (pr-str var)) {:reason :invalid-env-var}))
            {:var var
             :rel-path (normalize-relative-path rel)}))
        envpath-mappings))

(defn- split-kind-and-rest
  [line]
  (let [[_ kind rest] (re-matches #"^(\S+)\s+(.+)$" line)]
    [kind (some-> rest str/trim)]))

(defn- strip-inline-comment
  [line]
  (let [idx (.indexOf ^String line "#")]
    (if (<= idx 0)
      line
      (let [prefix (subs line 0 idx)
            prev (nth line (dec idx))]
        (if (and (not (str/blank? prefix)) (or (= prev \space) (= prev \tab)))
          (str/trim prefix)
          line)))))

(defn parse-string
  [s]
  (let [lines (str/split-lines (or s ""))]
    (loop [idx 0
           envs []
           files []
           envpaths []
           stdin nil]
      (if (>= idx (count lines))
        (let [request (parse-request envs files stdin)
              envpaths (parse-envpath-mappings envpaths)]
          {:request request
           :env-paths envpaths})
        (let [line-no (inc idx)
              raw (nth lines idx)
              line (-> raw str/trim strip-inline-comment str/trim)]
          (if (or (str/blank? line) (str/starts-with? line "#"))
            (recur (inc idx) envs files envpaths stdin)
            (let [[kind rest] (split-kind-and-rest line)]
              (when (or (str/blank? kind) (str/blank? rest))
                (invalid! (format "invalid map line %d: expected '<kind> <mapping>'" line-no)))
              (case kind
                "env" (recur (inc idx) (conj envs rest) files envpaths stdin)
                "file" (recur (inc idx) envs (conj files rest) envpaths stdin)
                "stdin"
                (do
                  (when stdin
                    (invalid! (format "invalid map line %d: multiple stdin entries" line-no)))
                  (recur (inc idx) envs files envpaths rest))
                "envpath" (recur (inc idx) envs files (conj envpaths rest) stdin)
                (invalid! (format "invalid map line %d: unknown kind %s" line-no (pr-str kind)))))))))))

(defn parse-file
  [path]
  (parse-string (slurp (io/file path))))

(defn resolve-profile
  [name]
  (let [name (some-> name str str/trim)]
    (when-not (valid-profile-name? name)
      (throw (ex-info (format "invalid profile name %s (allowed: letters, digits, ., _, -)" (pr-str name))
                      {:reason "invalid_profile_name"})))
    (let [filename (str name ".kmap")
          candidates (->> [(some-> (System/getenv env-profile-dir) str/trim not-empty)
                           (str (io/file ".kimen/profiles"))
                           (let [cfg-home (or (some-> (System/getenv "XDG_CONFIG_HOME") str/trim not-empty)
                                              (some-> (System/getProperty "user.home") str/trim not-empty (str "/.config")))]
                             (when cfg-home
                               (str (io/file cfg-home "kimen/profiles"))))]
                          (remove nil?)
                          (map #(str (io/file % filename)))
                          vec)
          hit (some (fn [p]
                      (when (.exists (io/file p))
                        p))
                    candidates)]
      (if hit
        hit
        (throw (ex-info (format "profile %s not found (set %s or create .kimen/profiles/%s)"
                                (pr-str name)
                                env-profile-dir
                                filename)
                        {:reason "invalid_profile_name"}))))))
